packages/plugin-ext/src/plugin/documents.ts

```diff
@@ -81,34 +81,70 @@ export class DocumentsExtImpl implements DocumentsExt {
-            if (data) {
-                const onWillSaveEvent: theia.TextDocumentWillSaveEvent = {
-                    document: data.document,
-                    reason: reason,
-                    /* tslint:disable:no-any */
-                    waitUntil: async (editsPromise: PromiseLike<theia.TextEdit[] | any>) => {
-                        const editsObjs = await editsPromise;
-                        if (this.isTextEditArray(editsObjs)) {
-                            const editOperations: SingleEditOperation[] = (editsObjs as theia.TextEdit[]).map(textEdit => Converter.fromTextEdit(textEdit));
-                            resolve(editOperations);
-                        } else {
-                            resolve([]);
-                        }
+
+    async $acceptModelWillSave(strUrl: UriComponents, reason: theia.TextDocumentSaveReason): Promise<SingleEditOperation[]> {
+        const uri = URI.revive(strUrl).toString();
+        const operations: SingleEditOperation[] = [];
+        let didTimeout = false;
+        const didTimeoutHandle = setTimeout(() => didTimeout = true, 1500);
+        try {
+            await this._onWillSaveTextDocument.sequence(async fireEvent => {
+                if (didTimeout) {
+                    return false;
+                }
+                try {
+                    const documentData = this.editorsAndDocuments.getDocument(uri);
+                    if (documentData) {
+                        const { document } = documentData;
+                        await this.fireTextDocumentWillSaveEvent({
+                            document, reason, fireEvent,
+                            accept: operation => operations.push(operation)
+                        });
                     }
-                };
-                this._onWillSaveTextDocument.fire(onWillSaveEvent);
-            }
-        });
+                } catch (e) {
+                    console.error(e);
+                }
+                return !didTimeout;
+            });
+        } finally {
+            clearTimeout(didTimeoutHandle);
+        }
+        return operations;
     }
 
-    isTextEditArray(obj: any): obj is theia.TextEdit[] {
-        return Array.isArray(obj) && obj.every((elem: any) => TextEdit.isTextEdit(elem));
+    // tslint:disable:no-any
+    protected async fireTextDocumentWillSaveEvent({
+        document, reason, fireEvent, accept
+    }: {
+            document: theia.TextDocument,
+            reason: theia.TextDocumentSaveReason,
+            fireEvent: (e: theia.TextDocumentWillSaveEvent) => any,
+            accept: (operation: SingleEditOperation) => void
+        }): Promise<void> {
+
+        const promises: PromiseLike<TextEdit[] | any>[] = [];
+        fireEvent(Object.freeze({
+            document, reason,
+            waitUntil(p: PromiseLike<TextEdit[] | any>) {
+                if (Object.isFrozen(promises)) {
+                    throw new Error('waitUntil can not be called async');
```

Comment:
Hello, I'm not sure the message will be meaningful for end users

Also it looks like a copy from vscode.

Reply:
End user should never see it. It's a programming error, extension developer should not call `waitUtil` async.

> Also it looks like a copy from vscode.

+1 will add copyrights

Reply:
actually code is copied from https://github.com/theia-ide/theia/blob/ad9e62bd78bf4da3ce829b25de64f9aa3fc8aaeb/packages/monaco/src/browser/monaco-editor-model.ts#L344-L366

only async is adjusted to be as an error in VS Code

not sure that we should open CQs for it